### DISCUSSION TITLE
help with convergence [likely BCs issue] of PorousFlow ThermoHydro hydrothermal model

### DISCUSSION THREAD

**Discussion by garciapintado** [idaholab/moose#21038] [Q&A Modules: General] [answered]
Hi all,
I am struggling to get convergence in a thermo-hydraulic model with PorousFlow. In my real case scenario, the model domain is a 2D cross-section of an ocean-basin, bounded at both sides by continental margins. The model aims to simulate basin-wide hydrothermal circulation. This is a density-driven circulation, where the thermal energy that heats the water comes form the underlying mantle. Thanks to you (discussion #20139, solved by @WilkAndy ) now PorousFlow is block-restricted, which has allowed me to go further by applying only non-PorousFlow thermal diffusion in the underlying mantle, so avoiding the high pressure and temperature subdomain, which were very problematic for the water EOS.
I can figure out some reasons by which the model still does not converge. Possibly, the tabulated extrapolation of the EOS [which I still need] is perhaps not god enough, but more than that I believe the top boundary conditions are probably not still in a good shape, which is driving the convergence process towards wildly sampling the EOS out of reasonable  P-T values.

To help me finding out the issue(s), I’ve been back to using box synthetic models of partial domains and playing with the boundary conditions. The attached image represents a model [after 6300 year simulation time] corresponding to the input file below. This model represents a block of porous material completely submerged [its top is at -3000m] under the ocean [other areas in my real model are sub-areal, in unsaturated conditions toward the surface]. That is, in this simple case it is assumed the model is top-bounded by ocean water, which can enter and exit freely the top of domain. No flow can go through the bottom, and flow should freely go in & out through the side boundaries [representing a continuation of porous matrix materials].

The bottom of the domain is [DirichletBC] at 800 ºC, representing a heat source from an upwelling melting body coming from the mantle. Porosity is constant, and permeability has a Gaussian-decay away from the center of the domain towards the sides, such that it allows for convection to develop only towards a centered column in the domain. 

I had hoped that this should still be an easy case for MOOSE-PorousFlow.  This synthetic model eventually converges, but still it goes through several steps relatively early in the simulation [~200 simulation years ] where it struggles to converge, and it samples out of the [very wide] tabulated range of P-T for the EOS. Overall I’d say the model is too slow [if I run it with a simple fluid, it is more than x100 times faster, but then the convection does not develop, so it is not realistic at all], and I am wondering if there is an issue with the BCs, which aggravates later on  in the “real-case” simulations [where the range of pressures and temperatures is wider, including unsaturated conditions] to the point of preventing convergence at all at the initial steps.

As you can see in the image, the hydrothermal plumes at the top of the domain have a natural shape, apparently leaving the domain without extraneous effects. To get this I’ve been playing with  PorousFlowPiecewiseLinearSink, for both pore pressure and temperature. By playing with the parameters for these BCs I’ve finally got hydrothermal plumes that do not tend to vanish nor to rebound when they reach the top of the domain. So, from a visualization perspective they look reasonable.

Still, being the convergence so problematic at several steps throughout the simulation, makes me think I am doing something wrong. The input file can be directly run with the standard MOOSE+PorousFlow+FluidProperties modules plus the attached fluid properties file. The input files and tabulated EOS it calls [I can upload the 2.2GB Exodus output file after 10000 yr if that helps; simulation is ~3h with 'mpiexec -np 4' my iMAC] is at https://github.com/garciapintado/MOOSE_porous_flow_stuff.git

Hope that you can help to spot  input errors and advise on ways to make the convergence go faster [as said, I am mostly suspicious of the BCs; and not even certain to be applying correctly the left and right ‘PorousFlowOutflowBC’ conditions, as the solution is not as symmetric as I would have expected regarding the lateral fluxes].

Best wishes,
Javier
—
![hydrothermal_synthetic00_6000yr](https://user-images.githubusercontent.com/44291802/168774375-23761c8f-d920-47c8-8f0e-c464a7d4dc76.png)

**Comment by WilkAndy** [comment]
Tagging @cpgr on this one.  I'm going on holiday soon so may be of limited use.

**Comment by cpgr** [comment, marked as answer]
Hi @garciapintado 

@WilkAndy knows I love these convection problems!

I think your main issue driving your convergence problem is that your maximum tilmestep size is too large. I set `dtmax = 1e9` arbitrarily, and each tilmestep converged in less than 5 nonlinear iterations and there was never a point where the solver tries to sample outside the specified pressure and temperature range. I also made a much smaller csv file of fluid properties to speed things up. Your one is 26 MB, so lots of entries. Most of the time required in  `TabulatedFluidProperties` is finding the correct position in the array for interpolation, so less points in your csv file = faster lookup. I'm not sure how you made yours, but I like to use the python CoolProp library.

Anyway, it looks like the warm water is escaping through the top, so your BC's look to be working for you. With the above changes, the following took about 45 minutes using 4 cores of my iMac. You could speed this up by coarsening the mesh in the upper regions and to the sides where nothing is happening (you could use less elements in the y direction and use the `bias_y` parameter, or even just refine in the middle section where everything happens).

https://user-images.githubusercontent.com/7664823/169177822-664a68ec-e23f-402e-a942-90d393c25a7a.mp4

PS: using `SimpleFluidProperties` doesn't lead to any convection, so all flow is diffusive which is why it runs so fast. The reason is that the density change over this temperature range is too small using a constant thermal expansion coefficient like `SimpleFluidProperties` uses in comparison to the real density change.

Also, here is the code I used to create a csv file of fluid properties
```python
import CoolProp.CoolProp as cp
import pandas as pd
import numpy as np

# Range of pressure and temperature
pressure = np.arange(1e7, 1.8e8+1, 5e6)
temperature = np.arange(273, 1300+1, 50)

data = np.zeros((pressure.size * temperature.size, 6))

for i in np.arange(pressure.size):
    for j in np.arange(temperature.size):
        data[i * temperature.size + j, 0] = pressure[i]
        data[i * temperature.size + j, 1] = temperature[j]
        data[i * temperature.size + j, 2] = cp.PropsSI('D', 'P', pressure[i], 'T', temperature[j], 'Water')
        data[i * temperature.size + j, 3] = cp.PropsSI('V', 'P', pressure[i], 'T', temperature[j], 'Water')
        data[i * temperature.size + j, 4] = cp.PropsSI('H', 'P', pressure[i], 'T', temperature[j], 'Water')
        data[i * temperature.size + j, 5] = cp.PropsSI('U', 'P', pressure[i], 'T', temperature[j], 'Water')

# Write out fluid_properties_extended.csv
df = pd.DataFrame(data, columns = ['pressure', 'temperature', 'density', 'viscosity', 'enthalpy', 'internal_energy'])
df.to_csv('fluid_properties_extended.csv', index=False)
```

**Reply by garciapintado** [reply]
Hi @cpgr,
Many thanks! This is extremely helpful and informative! I was unaware of CoolProp, and made an R script with a simple extrapolation towards extrema P,T values with a mild slope taken from the boundaries of the available P,T thermodynamic domain + a small mutiplicative factor.
I'll try your advices + script and will come with feedback [hopefully with the real domain problem; which has further issues]

**Reply by cpgr** [reply]
I forgot to mention that you could make your first time step much larger than `1`, as you give a good initial function for p and T. I used `1e4` I think (this saves a few time steps at the start). Also, you could increase the maximum tilmestep which would speed up the total compute time as well. I just guessed `1e9`s, you might get away with `3e9` or higher, just not `3e10`!

**Reply by garciapintado** [reply]
Thanks! A short question: why don't you get "cv" and "cp" for the tabulated fluid properties? And if so, where does MOOSE get these properties from?

**Reply by cpgr** [reply]
They aren't used in any of the kernels you have, so I didn't include them. If they were required, it would pass calculation back to the water fluid properties (in your case it would error out with a pressure out of range error).

**Reply by garciapintado** [reply]
Hi @cpgr,
I am using now your script based on CoolProp to get the thermodynamic fluid properties of water. Many thanks!
The limitation of the maximum timestep to 1e9 also works for me [higher values as 2e9 still lead to some sampling out of the validity range of the thermodynamic tables], and I am also using your discretisation steps for P,T so the file for the tabulated properties is much smaller.
Still, the simulation is not much faster on this side, and now and then I am getting some steps with take more than 10 nonlinear iterations to converge leading to smaller timesteps. All in all, it still takes 2.76 h [slightly less than before] in my iMAC, which is not that old [Processor: 3.6 GHz Intel Core i9; Memory: 64 GB2667 MHz DDR4] with ‘mpiexec -np 4´ processors,
and about the same time in a newer [1.5 years old] linux workstation.
I am wondering if it might be an issue with version of libraries [the last time I updated MOOSE+PETSc… was ~January in the iMAC]. In any case, I am going to try now a realistic case so I can be more specific about the real application at hand…

**Reply by cpgr** [reply]
Hmm, if you upload your new input file I can try and run it.

In any case, it is good that it is running.

**Reply by garciapintado** [reply]
I've uploaded it along with the python script to create the tabulated water properties [slightly wider range than your so I can use it for my 'real' case scenario, but still only 127KB size]
As it is, it has taken 9950 s in my iMAC with 'np -4'
I'm also close to finish an example with my real case scenario...

**Comment by garciapintado** [comment]
Hi @cpgr,

Assuming we’ll be able to figure out why the ’simple block’ simulation still is not as fast in my computer, in the meantime I’ve been back to my original, realistic case, scenario [on which @WilkAndy started to help earlier this year], here I am even more doubtful about my BCs [actually, pretty sure they are wrong!] but kind of lost about how to do it properly. Sorry, if you think this is a separate discussion. I can wait until the current one on the simple block is solved and then open another one with this message, but hoped this would give context on the nature of the real application.

For info, I am pasting an image resulting from a coupled  mechanical-hydrothermal simulation done with Matlab. I have done this is Matlab in a pretty raw way [splitting all operations], and I’hope to do it more properly in MOOSE, including some geochemistry [anhydrite deposition to close the pore space…]. I already have a mechanical code in Matlab, which by now the plan is to keep using it. The idea it that every 10000 years [the matlab mechanical model timestep] the matlab mesh [an evolving Lagrangian] would be exported to MOOSE, who would do the hydrothermal part and return the thermal state and the porepressure to Matlab. And so on. At least by now, I’d do the coupling just by system files forth and back every 10000 simulation years. The total coupled simulations would run for several millions years if we can afford that. The bottleneck is the hydrothermal part, which likely will run at a few years timesteps in MOOSE-PorousFlow.  So, the computing time it would take to do the hydrothermal part is critical to face the problem. The pasted image results from my current simple coupling between the mechanical and hydrothermal flow after ~10 Myr of hydrothermal activity [only the temperature feedback goes by now into the mechanical]. It shows the permeability field and the isotherms. The convection that develops around faults looks fine to me. I also believe that the hydrothermal activity extends too much towards the subaerial part, and don’t know how much of this is an artifact. In any case, the lack of geochemistry here [e.g. anhydrite deposition] seems to lead to unrealistically long-lived hydrothermal activity. My hope is to solve these issues in MOOSE+PorousFlow.
 
![siu3p nchem_hydr_melt_serp_spro COT sp3 ucr1000 066 001 01650 016500 000kyr ht2sp -250_250_-35_2x2 5_hydrothermal_permeability](https://user-images.githubusercontent.com/44291802/169525315-d8e252db-5b07-4785-abe3-fc7e54bcf94b.png)

<img width="1571" alt="Screenshot 2022-05-20 at 08 15 33" src="https://user-images.githubusercontent.com/44291802/169525687-f8df0760-32f4-4dc2-b519-330b6fb81bf9.png">

The second pasted image indicates the blocks I am using into MOOSE. So, in MOOSE, I am considering a lower block [not governed by PorousFlow], where only heat conduction applies [‘nopore’]. The upper blocks are governed by the PorousFlow dictator. The central part of the domain is submerged under the ocean [this covers the continental shelves; roughly the width of the ‘sediment’ area], while towards the sides, there is a subaerial part, where only barometric pressure should apply.  By now, I do not want to simulate the mechanical part in MOOSE as this is done in Matlab, so this would be a ThermoHydro coupling in PorousFlow. 

Also, I am not really interested in the unsaturated areas towards the top of subaerial part, but have to include them as well, as heat conduction should be considered in all the domain. To create the ‘unsat’ subdomain, I am simply making the assumption that a ‘phreatic level’ is above sea level by some 25% of the vertical distance between the topography and sea_level at each x-coordinate. My idea of doing so is to simply induce a mild flow from the sub-aerial part towards the submarine continental shelfs.

A problem I am facing now is how to set these top boundary conditions. Can this be done with a single-phase unsaturated simulation? If so, how would one set the porepressure boundaries on the top of the subaerial part so a mild inflow [kind of emulating rainfall] replenishes the subaerial part as it drains toward the ocean? 

A first alternative [a] I’ve considered is to say the ‘unsat’ block is also only subject to conduction. Then indicate some environment pressure and use PorousFlowPiecewiseLinearSink BCs for the porepressure at the interface between ‘unsat’ and ‘crust’. The flux allowed in this interface is very small, with the idea that it allows for some control of the porepressure at this interface but the small flux internal flux avoids some extraneous effects in the thermal field [i.e. I am not added the corresponding PorousFlowPiecewiseLinearSink for temperature at this inner interface, as conceptually I would not understand its meaning]. If I'm getting this right this would emulate a "confined" situation for the porous media under 'unsat', and the PorousFlowPiecewisewiseLinearSink would emulate the lack of the mechanical pressure, so continuously feeding the porepressure at this interface.
A simulation for 10000 year with a simple fluid takes only 671 secs [but thanks to you I understand this is because the lack of convection]. This develops some anomalies in temperature around the ‘unsat’ area, which I believe is because these internal BCs, but still Darcy fluxes close to the ocean bottom look to be in the right direction. However, if I try a real fluid [now with your CoolProp script] the simulation crashes very soon with messages as:

“Temperature 271.179 is outside the range of tabulated temperature (273.17, 1300).

A MooseException was raised during Auxiliary variable computation.
The next solve will fail, the timestep will be reduced, and we will try again.”

A second alternative [b], I’ve been considering is to consider ‘unsat’ as part of the PorousFlow domain, so this block is also subject to convection. In reality I’d favour this alternative if it worked. It is way slower than [a] for a simple fluid,  but positively the solution develops negative porepressures in the ‘unsat’ area, wich I understand is a good thing [still saturation is always 0, which I don’t understand]. Again, error messages similar to alternative [a] appear and the simulation crashes soon for the realistic EOS. Considering that ‘ptop_auxvar’ is the sum of a constant barometric pressure [1 atm] + the pressure by the corresponding water column, used as environment pressure, that ‘temperature_f’ is the temperature field imported from matlab, and that `top_submarine’ and ‘top_subaerial’ are respectively the nodesets under the ocean or over sea level, the BC block for [b] is as:


[BCs]
  [Ptop_submarine_sink]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = top_submarine               # nodes touching the ocean bottom
    PT_shift = ptop_auxvar                      # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    use_mobility = true
    use_relperm = true
    flux_function = 100
    fluid_phase = 0
    save_in = mass_flux_top_submarine_sink
  []
  [Ptop_submarine_source]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = top_submarine               # nodes touching the ocean bottom
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '-1e9 0'
    multipliers = '-1e9 0'
    use_mobility = true
    use_relperm = false
    flux_function = 100
    fluid_phase = 0
    save_in = mass_flux_top_submarine_source
  []
  [Ptop_subaerial_sink]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = top_subaerial               # nodes top-bounding the subDomain for porous flow simulation 
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    use_mobility = true
    use_relperm = true
    flux_function = 100 # similar to Ptop_submarine_sink to avoid shocks
    fluid_phase = 0
    save_in = mass_flux_top_subaerial_sink
  []
  [Ptop_subaerial_source]
    type = PorousFlowPiecewiseLinearSink  # 
    variable = porepressure
    boundary = top_subaerial               # nodes top-bounding the subDomain for porous flow simulation 
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '-1e9 0'
    multipliers = '-1e9 0'
    use_mobility = true
    use_relperm = false
    flux_function = 0.001 # to be calibrated: should allow for emptying of porespace
    fluid_phase = 0
    save_in = mass_flux_top_subaerial_source
  []
  [Ttop_submarine_sink]
    type = PorousFlowPiecewiseLinearSink
    variable = temperature
    boundary = top_submarine               # nodes touching the ocean bottom
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    use_mobility = true
    use_enthalpy = true
    use_relperm = true
    flux_function = 100
    fluid_phase = 0
    save_in = heat_flux_adv_top_submarine_sink
  []
  [Ttop_submarine_source]
    type = PorousFlowPiecewiseLinearSink
    variable = temperature
    boundary = top_submarine               # nodes touching the ocean bottom
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '-1e9 0'
    multipliers = '-1e9 0'
    use_mobility = true
    use_enthalpy = true
    use_relperm = false
    flux_function = 100
    fluid_phase = 0
    save_in = heat_flux_adv_top_submarine_source
  []
  [Ttop_subaerial_sink]
    type = PorousFlowPiecewiseLinearSink
    variable = temperature
    boundary = top_subaerial               # nodes top-bounding the subDomain for porous flow simulation 
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    use_mobility = true
    use_enthalpy = true
    use_relperm = true
    flux_function = 100 # similar to Ptop_submarine_sink to avoid shocks
    fluid_phase = 0
    save_in = heat_flux_adv_top_subaerial_sink
  []
  [Ttop_subaerial_source]
    type = PorousFlowPiecewiseLinearSink  # 
    variable = temperature
    boundary = top_subaerial               # nodes top-bounding the subDomain for porous flow simulation 
    PT_shift = ptop_auxvar                 # [Pa] reference environmental pressure: 1 atm + water column
    pt_vals = '-1e9 0'
    multipliers = '-1e9 0'
    use_mobility = true
    use_enthalpy = true
    use_relperm = false
    flux_function = 0.001 # to be calibrated: should allow for emptying of porespace
    fluid_phase = 0
    save_in = heat_flux_adv_top_subaerial_source
  []
  [Ttop_submarine_dif] # diffusive heat flux = - lambda * grad_T
    type = PorousFlowPiecewiseLinearSink
    variable = temperature
    PT_shift = 278.15                            # [K]    5ºC
    boundary = top_submarine
    pt_vals     = '-1000 0 1000'                  # x coordinates defining g
    multipliers = '-10000 0 1000'                 # y coordinates defining g
    use_thermal_conductivity = true 
    flux_function = 0.01
    save_in = heat_flux_dif_top_submarine
  []
  [Ttop_subaerial_dif] # diffusive heat flux = - lambda * grad_T
    type = PorousFlowPiecewiseLinearSink
    variable = temperature
    PT_shift = 278.15                            # [K]    5ºC
    boundary = top_subaerial
    pt_vals     = '-1000 0 1000'                  # x coordinates defining g
    multipliers = '-1000 0 1000'                 # y coordinates defining g
    use_thermal_conductivity = true 
    flux_function = 1
    save_in = heat_flux_dif_top_subaerial
  []
  [Tbot]
    type = FunctionDirichletBC 
    variable = temperature
    boundary = bottom
    function = temperature_f
  []
[]

I had expected that a low flux_function for the top_subaerial mass and advective heat BCs would lead to the side of the domain being slowly emptied. This seems to be the case as the popressure goes towards negative but the simple fluid, but this does not seem to happen with the tabulated EOS [although the latter crashes very soon, so maybe it is a matter of time]. Still, none of them has a saturation below 1 at any point, which is not what I expected. 

Darcy velocities across the ocean bottom look similar in both [a] and [b] for the simple fluid. For illustration I'm sending a plot comparing Darcy vertical velocities in both, still with the simple fluid after ~6000 years of simulation ([a] is the one on top):
<img width="1625" alt="Screenshot 2022-05-20 at 14 33 18" src="https://user-images.githubusercontent.com/44291802/169529178-7a981c1a-b586-474e-ad0a-9b96669e41b7.png">

In summary, I’m obviously doing things wrong and not understanding how these BCs interact with the model. Shall I upload to GitHub the input files so the simulation can be run if this helps you to help me? In the end I'd need to run this with the realsitic tabulated EOS.

I understand this is complicated and would not like to take much of your time. If there is something I could do to make your help not a big burden, but still get it please let me know.
Javier

**Reply by garciapintado** [reply]
Hi @cpgr
Just to note that this is for academic purposes in the context of marine environmental sciences [see #20139]. As I already mentioned to @WilkAndy. If we manage to have this running with your help, I will more than happy if eventually you were also interested in being a coauthor in ‘hopefully’ derived publications. I know this is off-topic now, but just wanted to mention.
Javier

**Reply by cpgr** [reply]
Hi Javier,

It would be great if you could upload this model to your GitHub so I can take a look at it. 

Chris

**Comment by garciapintado** [comment]
Hi Chris,
Many thanks! I’ve uploaded it to my Github, within the folder ‘rift_model’. For both, cases [a] and [b] the model converges with simple fluid, but crashes very early when using the tabulated EOS. I am also attaching the tabulated EOS file, which is the same as for the simple block experiment [corresponding to the script also in my Github].
Javier
---
On Tuesday, May 24, 2022 00:19 CEST, Chris Green ***@***.***> wrote:
 
 
Hi Javier,
It would be great if you could upload this model to your GitHub so I can take a look at it.
Chris

**Reply by cpgr** [reply]
Hi @garciapintado - I haven't forgotten about this (just been busy/away), but was wondering if you had a smaller problem that exhibits the same problem? Or do you think it would it be ok if I just cut this in half down a vertical line through the middle?

**Reply by garciapintado** [reply]
Hi @cpgr, Many thanks. I was wondering if it all looks too complicated. I haven't a smaller problem as this is derived from our realistic models of continental extension and rift generation towards oceanic spreading. A problem with coarser resolutions is that the faults do not develop so well, so we are stuck with this sort of problem size. Actually, our mesh in matlab has quadratic elements (a Tri6), and I am now importing it as Tri3 to MOOSE to make the problem faster and because in previous synthetic models with PorousFlow I found that in some circumstances that a Tri6 mesh did not converge while a Tri3 did! Also to make the problem easier to converge I am calculating by now the porosity and permeability fields in Matlab (approximately reproducing the PorousFlowPorosity ---simply cutting porosity below a threshold, rather than your non-negative porosity approach--- and KozenyCarman derived permeability formulations) and passing then as constant static fields to MOOSE (although eventually I'd like to do that also in MOOSE if everthing goes well; so they are updated along with the hydrothermal flow). About cutting in half down a vertical in the middle, I'd guess that if it works, it should work in all the domain. I need help but not sure how to help you to help me, and I understand that this application maybe is not so simple. If you think it'd be possible any useful, I could see to come to CSIRO for some weeks later after summer in Germany (after learning a bit more about MOOSE-PorousFlow).
